src/relcast.erl

```diff
@@ -998,11 +1041,88 @@ make_seq(ID, #state{seq_map=SeqMap, ids=A}=State) ->
 reset_seq(ID, #state{seq_map=SeqMap}=State) ->
     State#state{seq_map=maps:remove(ID, SeqMap)}.
 
-maybe_serialize(_Mod, Old, New, _Transaction) when Old == New ->
-    false;
-maybe_serialize(Mod, _Old, New, Transaction) ->
-    ok = rocksdb:transaction_put(Transaction, ?stored_module_state, Mod:serialize(New)),
-    true.
+maybe_serialize(_Mod, Old, #state{modulestate = New} = S) when Old == New ->
+    S;
+maybe_serialize(Mod, _Old, #state{modulestate = New0,
+                                  old_serialized = Old,
+                                  transaction = Transaction} = S) ->
+    New = Mod:serialize(New0),
+    KeyTree = do_serialize(Mod, Old, New, ?stored_key_prefix, Transaction),
+    S1 = maybe_write_key_tree(KeyTree, S),
+    S1#state{old_serialized = New, transaction_dirty = true}.
+
+maybe_write_key_tree(_KeyTree, #state{write_key_tree = false} = S) ->
+    S;
+maybe_write_key_tree(KeyTree, S) ->
+    ok = rocksdb:transaction_put(S#state.transaction, ?stored_key_tree,
+                                 term_to_binary(KeyTree, [compressed])),
+    S#state{write_key_tree = false}.
+
+do_serialize(Mod, Old, New, Prefix, Transaction) ->
+    case New of
+        State when is_binary(State) ->
+            ok = rocksdb:transaction_put(Transaction, ?stored_module_state, State);
+        StateMap ->
+            S = lists:sort(maps:to_list(StateMap)),
+            O = case Old of
```

Comment:
when the list lengths don't match, the state might have changed shape, and we need to account for that in upgrade/downgrade.